lang/java/reef-runtime-azbatch/src/main/java/org/apache/reef/runtime/azbatch/driver/AzureBatchEvaluatorShimManager.java

```diff
@@ -162,102 +186,135 @@ public void onNext(final RemoteMessage<EvaluatorShimProtocol.EvaluatorShimStatus
                                  final String command,
                                  final String evaluatorConfigurationString) {
-
-    String resourceRemoteId = this.activeResources.get(resourceLaunchEvent.getIdentifier());
-
-    EventHandler<EvaluatorShimProtocol.EvaluatorShimControlProto> handler = this.remoteManager
-        .getHandler(resourceRemoteId, EvaluatorShimProtocol.EvaluatorShimControlProto.class);
-
-    Set<FileResource> fileResources = resourceLaunchEvent.getFileSet();
-    String fileUrl = "";
-    if (!fileResources.isEmpty()) {
-      try {
-        File jarFile = writeFileResourcesJarFile(fileResources);
-        fileUrl = uploadFile(jarFile).toString();
-        LOG.log(Level.FINE, "Uploaded evaluator file resources to Azure Storage at {0}.", fileUrl);
-      } catch (IOException e) {
-        LOG.log(Level.SEVERE, "Failed to generate zip archive for evaluator file resources: {0}.", e);
-        throw new RuntimeException(e);
-      }
+    final String resourceId = resourceLaunchEvent.getIdentifier();
+
+    if (this.failedResources.containsKey(resourceId)) {
+      LOG.log(Level.FINE, "ResourceLaunch event triggered on a failed container. " +
+          "Notifying REEF of failed container.");
+      CloudTask cloudTask = this.failedResources.get(resourceId);
+      this.onAzureBatchTaskStatus(cloudTask);
+    } else if (this.evaluators.get(resourceId).isPresent()) {
+      LOG.log(Level.FINE, "Preparing to launch resourceId = {0}", resourceId);
+      this.launchEvaluator(resourceLaunchEvent, command, evaluatorConfigurationString);
     } else {
-      LOG.log(Level.INFO, "No file resources found in ResourceLaunchEvent.");
+      LOG.log(Level.WARNING, "Received a ResourceLaunch event for an unknown resourceId = {0}", resourceId);
     }
 
-    LOG.log(Level.INFO, "Sending a command to the Evaluator shim to start the evaluator.");
-    handler.onNext(
-        EvaluatorShimProtocol.EvaluatorShimControlProto
-            .newBuilder()
-            .setCommand(EvaluatorShimProtocol.EvaluatorShimCommand.LAUNCH_EVALUATOR)
-            .setEvaluatorLaunchCommand(command)
-            .setEvaluatorConfigString(evaluatorConfigurationString)
-            .setEvaluatorFileResourcesUrl(fileUrl)
-            .build());
-
     this.updateRuntimeStatus();
   }
 
+  /**
+   * Event handler method for {@link ResourceReleaseEvent}. Sends a TERMINATE command to the appropriate evaluator shim.
+   * @param resourceReleaseEvent
+   */
   public void onResourceReleased(final ResourceReleaseEvent resourceReleaseEvent) {
+    String resourceRemoteId = getResourceRemoteId(resourceReleaseEvent.getIdentifier());
+
+    // REEF Common will trigger a ResourceReleaseEvent even if the resource has failed. Since we know that the shim
+    // has already failed, we can safely ignore this.
+    if (this.failedResources.containsKey(resourceReleaseEvent.getIdentifier())) {
```

Comment:
Is concurrent access issue to failedResources here?

Reply:
Changed it to remove the element in line 277.